node/lib/util/push.js

```diff
@@ -66,38 +67,46 @@ const EMPTY_TREE = "4b825dc642cb6eb9a060e54bf8d69288fbee4904";
  *
  * @async
  * @param {NodeGit.Repository} repo
- * @param {String}             remoteName
  * @param {NodeGit.Commit}     commit
  */
-exports.getClosePushedCommit = co.wrap(function*(repo, remoteName, commit) {
+exports.getClosePushedCommit = co.wrap(function*(repo, commit) {
     assert.instanceOf(repo, NodeGit.Repository);
-    assert.isString(remoteName);
     assert.instanceOf(commit, NodeGit.Commit);
 
     // Search for the first commit that is not an ancestor of a remote ref, ie.
-    // the first commit that is unpushed. First by topological, then resolving
-    // by timestamp.
-
-    const walk = NodeGit.Revwalk.create(repo);
-    walk.sorting(NodeGit.Revwalk.SORT.TOPOLOGICAL,
-                 NodeGit.Revwalk.SORT.TIME,
-                 NodeGit.Revwalk.SORT.REVERSE);
-    walk.hideGlob(`refs/remotes/${remoteName}`);
-    walk.push(commit);
-
-    // This occurs when walk.next() has no commits left -- in this case,
-    // when there are no unpushed commits. Return the passed commit as the
-    // last pushed commit.
-
-    let firstNewOid;
-    try {
-        firstNewOid = yield walk.next();
-    } catch (err) {
-        if (NodeGit.Error.CODE.ITEROVER === err.errno)  {
-            return commit;
+    // the first commit that is unpushed. NodeGit Revwalk is too slow, so
+    // we shell out.
+
+    // First, get the list of remote refs to exclude
+    const excludeRefs = [];
+    for (const ref of (yield NodeGit.Reference.list(repo))) {
+        if (ref.startsWith("refs/remotes/")) {
+            excludeRefs.push("^" + ref);
         }
-        throw err;
     }
+    if (excludeRefs.length === 0) {
+        return null;
+    }
+
+    const execString = `\
+git -C '${repo.workdir()}' rev-list --reverse --topo-order ${commit} \
+${excludeRefs.join(" ")}`;
```

Comment:
do you want to limit the number of revs to return?  like `-n 1`?

Reply:
Yes, but unfortunately, I can't, because -n happens before --reverse.